trin-core/src/types/block_body.rs

```diff
@@ -0,0 +1,307 @@
+pub struct StorageKeys {
+    pub keys: Vec<H256>,
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use rstest::rstest;
+
+    // tx data from: https://etherscan.io/txs?block=14764013
+    #[rstest]
+    #[case(TX1, 102)]
+    #[case(TX2, 1078)]
+    #[case(TX3, 454)]
+    #[case(TX4, 455)]
+    #[case(TX5, 4414)]
+    #[case(TX6, 41942)]
+    #[case(TX7, 132984)]
+    #[case(TX8, 19177)]
+    #[case(TX9, 19178)]
+    #[case(TX10, 253851)]
+    #[case(TX11, 1908311)]
+    #[case(TX12, 233965)]
+    #[case(TX13, 6642)]
+    #[case(TX14, 3664042)]
+    #[case(TX15, 1544974)]
+    #[case(TX16, 3464424)]
+    #[case(TX17, 1544975)]
+    #[case(TX18, 12)]
+    fn encode_and_decode_txs(#[case] tx: &str, #[case] expected_nonce: u32) {
+        let tx_rlp = hex::decode(tx.to_string()).unwrap();
+        let tx: Transaction = Transaction::decode(&tx_rlp).expect("error decoding tx");
+        let expected_nonce = U256::from(expected_nonce);
+        match &tx {
+            Transaction::Legacy(tx) => assert_eq!(tx.nonce, expected_nonce),
+            Transaction::AccessList(tx) => assert_eq!(tx.nonce, expected_nonce),
+            Transaction::EIP1559(tx) => assert_eq!(tx.nonce, expected_nonce),
+        }
+        let encoded_tx = tx.encode();
+        assert_eq!(tx_rlp, encoded_tx);
+    }
+
+    #[test]
+    fn transactions_root() {
+        let tx_list: Vec<Transaction> = vec![
+            Transaction::decode(&hex::decode(TX1.to_string()).unwrap()).unwrap(),
+            Transaction::decode(&hex::decode(TX2.to_string()).unwrap()).unwrap(),
+            Transaction::decode(&hex::decode(TX3.to_string()).unwrap()).unwrap(),
+            Transaction::decode(&hex::decode(TX4.to_string()).unwrap()).unwrap(),
+            Transaction::decode(&hex::decode(TX5.to_string()).unwrap()).unwrap(),
+            Transaction::decode(&hex::decode(TX6.to_string()).unwrap()).unwrap(),
+            Transaction::decode(&hex::decode(TX7.to_string()).unwrap()).unwrap(),
+            Transaction::decode(&hex::decode(TX8.to_string()).unwrap()).unwrap(),
+            Transaction::decode(&hex::decode(TX9.to_string()).unwrap()).unwrap(),
+            Transaction::decode(&hex::decode(TX10.to_string()).unwrap()).unwrap(),
+            Transaction::decode(&hex::decode(TX11.to_string()).unwrap()).unwrap(),
+            Transaction::decode(&hex::decode(TX12.to_string()).unwrap()).unwrap(),
+            Transaction::decode(&hex::decode(TX13.to_string()).unwrap()).unwrap(),
+            Transaction::decode(&hex::decode(TX14.to_string()).unwrap()).unwrap(),
+            Transaction::decode(&hex::decode(TX15.to_string()).unwrap()).unwrap(),
+            Transaction::decode(&hex::decode(TX16.to_string()).unwrap()).unwrap(),
+            Transaction::decode(&hex::decode(TX17.to_string()).unwrap()).unwrap(),
+            Transaction::decode(&hex::decode(TX18.to_string()).unwrap()).unwrap(),
+            Transaction::decode(&hex::decode(TX19.to_string()).unwrap()).unwrap(),
```

Comment:
Not a big deal, but it was surprising that this list went to 19, and the earlier list went to 18. Makes me wonder if something is wrong...

Reply:
Ahh, I just forgot to include tx 19 on the first test - but I've added it now- which simply tests that each tx individually decodes properly (maybe somewhat redundant to have that many individual txs tested?). The latter test is what's testing proper calculation of the trie root - if I remove a tx from that list the test will fail.

Reply:
Yeah, I think it would be fine to have fewer tests in the first test, it just made me cringe to have all but one 😆 .